Rust gives wrong help. (&mut mut Vec<T>)

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

So this is the minimal code:

<details>

```rust
fn main() {
    let mut test = Vec::new();

    let rofl: &Vec<Vec<i32>> = &mut test;

    rofl.push(Vec::new());
}
```

</details>

If I'm running `cargo check` on this, than I'm getting the following:

<details>

```rust
    Checking rust_tmp v0.1.0 (/tmp/rust_tmp)
error[E0596]: cannot borrow `*rofl` as mutable, as it is behind a `&` reference
 --> src/main.rs:6:5
  |
4 |     let rofl: &Vec<Vec<i32>> = &mut test;
  |                                --------- help: consider changing this to be a mutable reference: `&mut mut test`
5 |
6 |     rofl.push(Vec::new());
  |     ^^^^ `rofl` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error: aborting due to previous error

For more information about this error, try `rustc --explain E0596`.
error: could not compile `rust_tmp`

To learn more, run the command again with --verbose.

[Process exited 101]
```

</details>

So the interesting part is this one:
```rust
  |   --------- help: consider changing this to be a mutable reference: `&mut mut test`
```

If I try this:

<details>

```rust
fn main() {
    let mut test = Vec::new();

    // Notice the second "mut" after the "="
    let rofl: &Vec<Vec<i32>> = &mut mut test;

    rofl.push(Vec::new());
}
```

</details>

Than I'm getting this when running `cargo check`:

<details>

```rust
    Checking rust_tmp v0.1.0 (/tmp/rust_tmp)
error: expected expression, found keyword `mut`
 --> src/main.rs:5:37
  |
5 |     let rofl: &Vec<Vec<i32>> = &mut mut test;
  |                                     ^^^ expected expression

error: aborting due to previous error

error: could not compile `rust_tmp`

To learn more, run the command again with --verbose.

[Process exited 101]
```

</details>

Well I expected that `rust` would say, that the given type of `rofl` is wrong or doesn't fit with the value which is gonna be assigned. Something like this:

<details>

```rust
    Checking rust_tmp v0.1.0 (/tmp/rust_tmp)
error[E0596]: cannot borrow `*rofl` as mutable, as it is behind a `&` reference
 --> src/main.rs:7:5
  |
5 |     let rofl: &Vec<Vec<i32>> = &mut test;
  |               --- help: consider changing this to be a mutable reference:
  |                         `&mut Vec<Vec<i32>>`
6 |
7 |     rofl.push(Vec::new());
  |     ^^^^ `rofl` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error: aborting due to previous error

For more information about this error, try `rustc --explain E0596`.
error: could not compile `rust_tmp`

To learn more, run the command again with --verbose.

[Process exited 101]
```

</details>

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:

<details>

```
rustc 1.54.0-nightly (5c0292654 2021-05-11)
binary: rustc
commit-hash: 5c029265465301fe9cb3960ce2a5da6c99b8dcf2
commit-date: 2021-05-11
host: x86_64-unknown-linux-gnu
release: 1.54.0-nightly
LLVM version: 12.0.1
```

</details>

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary>Backtrace</summary>
<p>

It didn't even build, so I think it's fine if I don't fill this part.

</p>
</details>


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"fee1-dead"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs
+++ b/compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs
@@ -902,9 +902,13 @@ fn suggest_ampmut<'tcx>(
             {
                 let lt_name = &src[1..ws_pos];
                 let ty = &src[ws_pos..];
-                return (assignment_rhs_span, format!("&{} mut {}", lt_name, ty));
+                if !ty.trim_start().starts_with("mut") {
+                    return (assignment_rhs_span, format!("&{} mut {}", lt_name, ty));
+                }
             } else if let Some(stripped) = src.strip_prefix('&') {
-                return (assignment_rhs_span, format!("&mut {}", stripped));
+                if !stripped.trim_start().starts_with("mut") {
+                    return (assignment_rhs_span, format!("&mut {}", stripped));
+                }
             }
         }
     }

--- a/src/test/ui/borrowck/issue-85765.rs
+++ b/src/test/ui/borrowck/issue-85765.rs
@@ -1,0 +1,8 @@
+fn main() {
+    let mut test = Vec::new();
+    let rofl: &Vec<Vec<i32>> = &mut test;
+    //~^ HELP consider changing this to be a mutable reference
+    rofl.push(Vec::new());
+    //~^ ERROR cannot borrow `*rofl` as mutable, as it is behind a `&` reference
+    //~| NOTE `rofl` is a `&` reference, so the data it refers to cannot be borrowed as mutable
+}

--- a/src/test/ui/borrowck/issue-85765.stderr
+++ b/src/test/ui/borrowck/issue-85765.stderr
@@ -1,0 +1,12 @@
+error[E0596]: cannot borrow `*rofl` as mutable, as it is behind a `&` reference
+  --> $DIR/issue-85765.rs:5:5
+   |
+LL |     let rofl: &Vec<Vec<i32>> = &mut test;
+   |         ---- help: consider changing this to be a mutable reference: `&mut Vec<Vec<i32>>`
+LL |
+LL |     rofl.push(Vec::new());
+   |     ^^^^ `rofl` is a `&` reference, so the data it refers to cannot be borrowed as mutable
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0596`.

